src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java

```diff
@@ -410,7 +410,7 @@ private int implWrite(byte[] b, int off, int len) throws IOException {
         } catch (InterruptedIOException e) {
             throw e;
         } catch (IOException ioe) {
-            throw new SocketException(ioe.getMessage());
+            throw new SocketException(ioe.getMessage(), ioe);
```

Comment:
I'm in two minds on whether this is a good idea or not because the stack trace of the cause is mostly the same as the SocketException, e.g. right now we get a clear stack trace like this:
```
Exception in thread "main" java.net.SocketException: Broken pipe
	at java.base/sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:413)
	at java.base/sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:433)
	at java.base/sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:812)
	at java.base/java.net.Socket$SocketOutputStream.write(Socket.java:1192)
	at java.base/java.io.OutputStream.write(OutputStream.java:124)
```

but with the change we get a cause with almost the same stack trace:
```
Exception in thread "main" java.net.SocketException: Broken pipe
	at java.base/sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:413)
	at java.base/sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:433)
	at java.base/sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:812)
	at java.base/java.net.Socket$SocketOutputStream.write(Socket.java:1192)
	at java.base/java.io.OutputStream.write(OutputStream.java:124)
Caused by: java.io.IOException: Broken pipe
	at java.base/sun.nio.ch.SocketDispatcher.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:62)
	at java.base/sun.nio.ch.NioSocketImpl.tryWrite(NioSocketImpl.java:388)
	at java.base/sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:404)
	at java.base/sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:433)
	at java.base/sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:812)
	at java.base/java.net.Socket$SocketOutputStream.write(Socket.java:1192)
	at java.base/java.io.OutputStream.write(OutputStream.java:124)
```

The reason for the translation from IOException to SocketException is so that the SocketImpl added via JEP 353 throws the same specific exceptions as the legacy implementation. We were concerned there may be older code assuming that read/write throw SocketException. So one idea here is to avoid the confusing cause is to use the stack trace from the IOException, as in:
```
var e = new SocketException(ioe.getMessage());
e.setStackTrace(ioe.getStackTrace());
throw e;
```
which gives us a clear stack trace like we have now:
```
Exception in thread "main" java.net.SocketException: Broken pipe
	at java.base/sun.nio.ch.SocketDispatcher.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:62)
	at java.base/sun.nio.ch.NioSocketImpl.tryWrite(NioSocketImpl.java:388)
	at java.base/sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:404)
	at java.base/sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:435)
	at java.base/sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:814)
	at java.base/java.net.Socket$SocketOutputStream.write(Socket.java:1192)
	at java.base/java.io.OutputStream.write(OutputStream.java:124)
```

Reply:
Hi Alan, I adjusted this coding following your suggestion to use setStackTrace.

Reply:
> Hi Alan, I adjusted this coding following your suggestion to use setStackTrace.

That's just one of them. In a later comment I pasted in asSocketException and changed all 4 cases to use it, maybe you could try that. The right thing is of course to drop the translation to SocketException as anything using read/write has to deal with IOException anyway, it's really just a concern that older code may having special handling of SocketException of undocumented, but longstanding, behavior in older JDK releases.

Reply:
Sorry Alan, seems I somehow missed that. I pushed a new commit including asSocketException  .


I wonder, could asSocketException  in the form of asOtherException maybe be a useful generics that covers multiple exception classes (but that would be something for another PR) ?
```
public static <T extends Exception> T asOtherException(Exception e, T n) {
    ...
}
```
(maybe we could omit the second param somehow, not sure.)

Reply:
If there were other translations needed then it might be useful but I don't think we need it for this change.  So I think go with what you have now and we can re-visit it another time if needed, hopefully to remove it (at least for read/write) as the translation shouldn't be required.